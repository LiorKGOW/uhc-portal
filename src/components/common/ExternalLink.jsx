import React from 'react';
import PropTypes from 'prop-types';
import { ExternalLinkAltIcon } from '@patternfly/react-icons';
import { trackEvents } from '~/common/analytics';
import useAnalytics from '~/hooks/useAnalytics';

const ExternalLink = ({ href, children, noIcon, noTarget, className, stopClickPropagation }) => {
  const track = useAnalytics();

  const trackExternalLink = () => {
    const path = window.location.pathname;
    let resource;
    if (path.includes('/rosa')) {
      resource = 'moa';
    } else if (path.includes('/osdtrial')) {
      resource = 'osdtrial';
    } else if (path.includes('/osd')) {
      resource = 'osd';
    } else if (path.includes('/crc')) {
      resource = 'crc';
    } else {
      resource = 'all';
    }

    track(trackEvents.ExternalLink, {
      customProperties: {
        link_url: href,
        module: 'openshift',
        ocm_resource_type: resource,
      },
    });
  };

  const handleClick = (event) => {
    if (stopClickPropagation) {
      trackExternalLink();
      event.stopPropagation();
    } else {
      trackExternalLink();
    }
  };

  return (
    <a
      href={href}
      target={noTarget ? '' : '_blank'}
      rel="noreferrer noopener"
      className={className}
      onClick={handleClick}
    >
      {children}
      {noTarget ? null : <span className="pf-u-screen-reader"> (new window or tab)</span>}
      {!noIcon && <ExternalLinkAltIcon color="#0066cc" size="sm" className="pf-u-ml-sm" />}
    </a>
  );
};

ExternalLink.propTypes = {
  href: PropTypes.string.isRequired,
  children: PropTypes.node,
  noIcon: PropTypes.bool,
  noTarget: PropTypes.bool,
  className: PropTypes.string,
  stopClickPropagation: PropTypes.bool,
};

export default ExternalLink;
