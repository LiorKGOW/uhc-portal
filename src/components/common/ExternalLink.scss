.external-link-alt-icon {
    margin-left: 3px !important; //.pf-v5-u-ml-sm has a !important property, so...
    margin-right: 2px;
}