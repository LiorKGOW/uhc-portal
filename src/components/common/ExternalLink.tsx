import React from 'react';
import { ExternalLinkAltIcon } from '@patternfly/react-icons';
import { trackEvents } from '~/common/analytics';
import useAnalytics from '~/hooks/useAnalytics';

const ExternalLink = ({ href, children, noIcon, noTarget, className }: ExternalLinkProps) => {
  const { track } = useAnalytics();
  const userUrl = window.location.href;

  //todo
  // module: window._segment?.activeModule,
  const module = 'test module';

  const resourceType = () => {
    if (userUrl.includes('/rosa')) {
      return 'moa';
    } else if (userUrl.includes('/osdtrial')) {
      return 'osdtrial';
    } else if (userUrl.includes('/osd')) {
      return 'osd';
    } else if (userUrl.includes('/crc')) {
      return 'crc';
    } else {
      return 'all';
    }
  };

  const trackExternalLink = () => {
    track(trackEvents.ExternalLink, {
      customProperties: JSON.parse(
        `{"link_url":"${href}", "module":"${module}", "ocm_resource_type":"${resourceType()}"}`,
      ),
    });
  };

  return (
    <a
      href={href}
      target={noTarget ? '' : '_blank'}
      rel="noreferrer noopener"
      className={className}
      onClick={trackExternalLink}
    >
      {children}
      {noTarget ? null : <span className="pf-u-screen-reader"> (new window or tab)</span>}
      {!noIcon && <ExternalLinkAltIcon color="#0066CC" size="sm" className="pf-u-ml-sm" />}
    </a>
  );
};

interface ExternalLinkProps {
  href: string;
  children?: React.ReactNode;
  noIcon?: boolean;
  noTarget?: boolean;
  className?: string;
}

export default ExternalLink;
