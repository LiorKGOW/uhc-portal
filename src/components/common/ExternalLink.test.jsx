import React from 'react';
import { render, screen } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import '@testing-library/jest-dom';

import ExternalLink from './ExternalLink';

expect.extend(toHaveNoViolations);

describe('<ExternalLink />', () => {
  it('should render with basic options', async () => {
    const { container } = render(
      <ExternalLink href="http://example.com">Hello World</ExternalLink>,
    );
    expect(container).toMatchSnapshot();

    const results = await axe(container);
    expect(results).toHaveNoViolations();

    expect(screen.getByText('(new window or tab)')).toBeInTheDocument();
    expect(container.querySelector('svg')).toBeInTheDocument();
    expect(container.querySelector('a')).toBeInTheDocument();
  });

  it('does not display an icon if "noIcon" is enabled', () => {
    const { container } = render(
      <ExternalLink href="http://example.com" noIcon>
        Hello World
      </ExternalLink>,
    );

    expect(screen.getByText('(new window or tab)')).toBeInTheDocument();
    expect(container.querySelector('svg')).not.toBeInTheDocument();
    expect(container.querySelector('.pf-c-button')).not.toBeInTheDocument();
  });

  it('does does not set a target if "noTarget" is enabled ', () => {
    const { container } = render(
      <ExternalLink href="http://example.com" noTarget>
        Hello World
      </ExternalLink>,
    );

    expect(screen.queryByText('(new window or tab)')).not.toBeInTheDocument();
    expect(container.querySelector('svg')).toBeInTheDocument();
    expect(container.querySelector('.pf-c-button')).not.toBeInTheDocument();
  });

  it('displays as a button if "isButton" is enabled', () => {
    const { container } = render(
      <ExternalLink href="http://example.com" isButton>
        Hello World
      </ExternalLink>,
    );

    expect(screen.getByText('(new window or tab)')).toBeInTheDocument();
    expect(container.querySelector('svg')).toBeInTheDocument();
    expect(container.querySelector('.pf-c-button')).toBeInTheDocument();
  });
});
