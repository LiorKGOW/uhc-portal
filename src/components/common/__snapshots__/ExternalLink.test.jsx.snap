// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExternalLink /> should render 1`] = `
<a
  href="http://example.com"
  rel="noreferrer noopener"
  target="_blank"
>
  Hello World
  <span
    className="pf-u-screen-reader"
  >
     (new window or tab)
  </span>
   
  <ExternalLinkAltIcon
    color="#0066cc"
    noVerticalAlign={false}
    size="sm"
  />
</a>
`;
